worker/src/main/java/org/apache/celeborn/service/deploy/worker/storage/FileWriter.java

```diff
@@ -152,25 +152,26 @@ public void decrementPendingWrites() {
     numPendingWrites.decrementAndGet();
   }
 
-  @GuardedBy("flushLock")
   protected void flush(boolean finalFlush) throws IOException {
-    // flushBuffer == null here means writer already closed
-    if (flushBuffer != null) {
-      int numBytes = flushBuffer.readableBytes();
-      if (numBytes != 0) {
-        notifier.checkException();
-        notifier.numPendingFlushes.incrementAndGet();
-        FlushTask task = null;
-        if (channel != null) {
-          task = new LocalFlushTask(flushBuffer, channel, notifier);
-        } else if (fileInfo.isHdfs()) {
-          task = new HdfsFlushTask(flushBuffer, fileInfo.getHdfsPath(), notifier);
-        }
-        addTask(task);
-        flushBuffer = null;
-        fileInfo.updateBytesFlushed(numBytes);
-        if (!finalFlush) {
-          takeBuffer();
+    synchronized (flushLock) {
```

Comment:
There are a few locations where `flushLock` is not held when invoking `flush` - which is why error-prone ended up flagging it.
For example:
* `ReducePartitionFileWriter.flush`
* Test cases.

Reply:
Yeah, I checked locations where `flush` is invoked without synchronizing `flushLock` and confirmed that there wasn't concurrent bug, but I DO agree with your refactor since it's more robust against future evolution :)